Create Apache Datafusion upgrade issue template (#6800)
spiceai/spiceai: .github/ISSUE_TEMPLATE/datafusion_upgrade.md

---

> name: Apache DataFusion Upgrade

Reviewer: ```suggestion
name: DataFusion Upgrade
```

No need for the Apache prefix

about: Checklist for upgrading to a new major version of Apache DataFusion.

> title: 'Upgrade Apache DataFusion to version X.Y.Z'

Reviewer: ```suggestion
about: Checklist for upgrading to a new major version of DataFusion.
title: 'Upgrade DataFusion to version X'
```

labels: enhancement
assignees: ''

---

> This issue tracks the process of upgrading Spice OSS to a new major version of Apache DataFusion. We endeavor to maintain a version that is one major release version behind the [latest](https://github.com/apache/datafusion/tags). This is a complex process because many of our internal crates and forked dependencies rely on DataFusion. Ensuring they all use the same version is critical for stability and performance.

Reviewer 2: ```suggestion
This issue tracks the process of upgrading Spice OSS to a new major version of Apache DataFusion. Maintain a version that is one major release version behind the [latest](https://github.com/apache/datafusion/tags). This is a complex process because many of Spice's internal crates and forked dependencies rely on DataFusion. Ensuring they all use the same version is critical for stability and performance.
```

> This issue tracks the process of upgrading Spice OSS to a new major version of Apache DataFusion. We endeavor to maintain a version that is one major release version behind the [latest](https://github.com/apache/datafusion/tags). This is a complex process because many of our internal crates and forked dependencies rely on DataFusion. Ensuring they all use the same version is critical for stability and performance.

Reviewer: ```suggestion
This issue tracks upgrading Spice OSS to a new major version of Apache DataFusion. The project targets staying one major release behind the [latest](https://github.com/apache/datafusion/tags). Because many internal crates and forked dependencies rely on DataFusion, they all need to be upgraded in lockstep.
```

> ## Pre-upgrade Tasks

Reviewer: DataFusion has also started publishing upgrade guides: https://datafusion.apache.org/library-user-guide/upgrading.html and blogs: https://datafusion.apache.org/blog/2025/07/28/datafusion-49.0.0/ that should be required reading.

- [ ]  Read the Apache DataFusion [changelog](https://github.com/apache/datafusion/tree/branch-49/dev/changelog) of the new version to identify breaking changes and new features.
- [ ]  Create a new branch for the upgrade process.

> ## Forked Dependency Upgrades

Reviewer: There is a broader set of crates that needs to be updated when a new DF version also brings a new arrow-rs major version, I think this includes (but may not be limited to):
- snowflake-rs
- delta-kernel-rs
- duckdb-rs
- maybe others?

> The following forked dependencies use DataFusion and need to be upgraded to maintain compatibility. This typically involves pulling the latest changes from the upstream repository, resolving conflicts, and updating the commit hash in our `Cargo.toml` (See [Core Dependency Upgrade](#core-dependency-upgrade)).

Reviewer: ```suggestion
The following forked dependencies use DataFusion and need to be upgraded in lockstep. This typically involves pulling the latest changes from the upstream repository, resolving conflicts, and updating the commit hash in `Cargo.toml` (See [Core Dependency Upgrade](#core-dependency-upgrade)).
```

> The following forked dependencies use DataFusion and need to be upgraded to maintain compatibility. This typically involves pulling the latest changes from the upstream repository, resolving conflicts, and updating the commit hash in our `Cargo.toml` (See [Core Dependency Upgrade](#core-dependency-upgrade)).

Reviewer: We should also evaluate whether we still need our fork - sometimes we can get our changes upstreamed and into the next version, so we can switch back to the published crates.io version.

> - [ ]  **datafusion**: Update our fork of the datafusion repo.

Reviewer: ```suggestion
- [ ]  **datafusion**: Update the spiceai fork of the datafusion repo.
```

> - [ ]  **datafusion**: Update our fork of the datafusion repo.

Reviewer: We should also have an item to review all commits that are not upstreamed and ping the authors of the those commits to get their changes upstreamed (if they haven't yet). We may never get there, but we should aim for 0 changes from datafusion.

>   - [ ] Branch from the lastest `spice-<X-1>` branch and name it `spice-<X>`. 

Reviewer: I noticed we deviated this time, but I think its cleaner to branch from the DataFusion tag, and then apply our commits from the previous branch on top of the new branch. This does make it harder to upgrade, but it makes it much easier to see which changes have yet to be upstreamed (and adds more motivation to ensure our changes get upstreamed).

  - [ ] Merge from the upstream release tag `vX.Y.Z`

> - [ ]  **datafusion-federation**: Update our fork to be compatible with the new DataFusion version.

Reviewer: For each crate, let's list the branching strategy we use. For datafusion-federation, we branch off the previous `spiceai-<dfX>` and create `spiceai-<dfX+1>` and add the changes into there.

> - [ ]  **datafusion-table-providers**: Update our fork to be compatible with the new DataFusion version.

Reviewer: `datafusion-table-providers` has a long-lived `spiceai` branch. The DF upgrade should not be merged until the main spiceai PR is ready to be merged - as it would block any other PRs from going in on top of the DF upgrade PR.

Reviewer: We deviated from this and now William's PR https://github.com/datafusion-contrib/datafusion-table-providers/pull/420 is blocked from being merged into spiceai until the spiceai PR to upgrade DF is merged.

- [ ]  **iceberg-rust**: The `iceberg-datafusion` crate within this forked repository needs to be updated.

> 

Reviewer 2: Add links to the forks here maybe?

> 

Reviewer 2: ```suggestion
- [ ]  **datafusion**: Update the fork of the datafusion repo.
  - [ ] Branch from the lastest `spice-<X-1>` branch and name it `spice-<X>`. 
  - [ ] Merge from the upstream release tag `vX.Y.Z`
- [ ]  **datafusion-federation**: Update the fork to be compatible with the new DataFusion version.
- [ ]  **datafusion-table-providers**: Update the fork to be compatible with the new DataFusion version.
- [ ]  **iceberg-rust**: The `iceberg-datafusion` crate within this forked repository needs to be updated.
```

## Core Dependency Upgrade

- [ ]  Update the `datafusion` dependency in the root `Cargo.toml` to the new patched commit.

> - [ ]  Update the `arrow-rs` dependency in the root `Cargo.toml` to the new patched commit.

Reviewer: `arrow-rs` shouldn't need to be updated? Only if the new DF version brings with it a new arrow-rs major version bump.

Reviewer: maybe have a separate section for the arrow-rs specific upgrades (i.e. including crates).

- [ ]  Update the `datafusion-federation` dependency in the root `Cargo.toml` to the new patched commit.
- [ ]  Update the `datafusion-table-providers` dependency in the root `Cargo.toml` to the new patched commit.
- [ ]  Update the `datafusion-federation` dependency in the root `Cargo.toml` to the new patched commit.

## Build and Test

> - [ ]  Run `make build` to ensure the entire project compiles without errors.

Reviewer: We need to run all of the testoperator benchmark tests and update snapshots if required. (This should happen before the upgrade PR is merged)

- [ ]  Run all tests using `make build-cli nextest` to verify that all functionality is working as expected and snapshots have not changed.
- [ ]  Address any compilation errors or test failures. This may involve fixing code that is incompatible with the new DataFusion version.

## Post-upgrade Tasks

- [ ]  Create a pull request with the changes.
- [ ]  Ensure all CI checks pass.

> - [ ]  Build the branch version and test with test operator, updating snapshots if needed.

Reviewer 3: I would move this pre-upgrade. We would want to run testoperator with query validation prior to merging the upgrade.

The following forked dependencies use DataFusion and need to be upgraded in lockstep. This typically involves pulling the latest changes from the upstream repository, resolving conflicts, and updating the commit hash in `Cargo.toml` (See [Core Dependency Upgrade](#core-dependency-upgrade)).
- [ ]  **[datafusion-federation](https://github.com/spiceai/datafusion-federation)**: Update the fork to be compatible with the new DataFusion version.
- [ ]  **[datafusion-table-providers](https://github.com/datafusion-contrib/datafusion-table-providers)**: Update the fork to be compatible with the new DataFusion version.
  - Do not merge the into the `spiceai` branch until the main Spice OSS PR is ready to be merged. Merging sooner can block other PRs.
- [ ]  **[iceberg-rust](https://github.com/spiceai/iceberg-rust.git)**: The `iceberg-datafusion` crate within this forked repository needs to be updated.

## Arrow Updates (if necessary)
Spice should use the same version of Arrow that DataFusion uses. If DataFusion upgraded Arrow, then the following crates should be upgraded.

- [ ]  **arrow-rs**: 
- [ ]  **snowflake-rs**: 
- [ ]  **delta-kernel-rs**: 
- [ ]  **duckdb-rs**: 

> ## Core Dependency Upgrade

Reviewer 4: We can probably recommend to use own personal branches until tests are completed and we are fully ready to proceed with merge
 - avoid potential issues with protected branches
 - avoid potential issues with locking currently active branch (unable to add changes until in-progress upgrade is completed), like datafusion-table-providers/spiceai
